rclcpp_lifecycle/src/state.cpp

```diff
@@ -86,30 +86,33 @@ State::~State()
 State &
 State::operator=(const State & rhs)
 {
-  if (this != &rhs) {
-    reset();
+  if (this == &rhs) {
+    return *this;
+  }
+
+  // reset all currently used resources
+  reset();
+
+  allocator_ = rhs.allocator_;
+  owns_rcl_state_handle_ = rhs.owns_rcl_state_handle_;
+
+  // we don't own the handle, so we can return straight ahead
+  if (!owns_rcl_state_handle_) {
+    state_handle_ = rhs.state_handle_;
+    return *this;
+  }
 
-    allocator_ = rhs.allocator_;
-    owns_rcl_state_handle_ = rhs.owns_rcl_state_handle_;
-
-    if (!owns_rcl_state_handle_) {
-      state_handle_ = rhs.state_handle_;
-    } else {
-      state_handle_ = static_cast<rcl_lifecycle_state_t *>(
-        allocator_.allocate(sizeof(rcl_lifecycle_state_t), allocator_.state));
-      if (!state_handle_) {
-        throw std::runtime_error("failed to allocate memory for rcl_lifecycle_state_t");
-      }
-      state_handle_->id = rhs.state_handle_->id;
-      state_handle_->label = rcutils_strndup(
-        rhs.state_handle_->label,
-        strlen(rhs.state_handle_->label),
-        allocator_);
-      if (!state_handle_->label) {
-        reset();
-        throw std::runtime_error("failed to duplicate label for rcl_lifecycle_state_t");
-      }
-    }
+  // we own the handle, so we have to deep-copy the rhs object
+  state_handle_ = static_cast<rcl_lifecycle_state_t *>(
+    allocator_.allocate(sizeof(rcl_lifecycle_state_t), allocator_.state));
+  if (!state_handle_) {
+    throw std::runtime_error("failed to allocate memory for rcl_lifecycle_state_t");
+  }
+  auto ret = rcl_lifecycle_state_init(
+    state_handle_, rhs.id(), rhs.label().c_str(), &allocator_);
+  if (ret != RCL_RET_OK) {
+    reset();
```

Comment:
Again, `state_handle_` has not been zero initialized, so no guarantee that `reset` can do the right thing.

Same below.

Reply:
I don't see how. If the init returns false, it means that the label pointer is null and thus can be freed correctly.
The same with transitions. We call init on it with two nullptr.